ValkeyLDAP.md

```diff
@@ -0,0 +1,271 @@
+- Audit trail for authentication events
+- TLS 1.2+ required for AD communication
+- Certificate validation
+- Secure credential storage
+- Encrypted cache entries
+
+### Maintainability
+
+- Clear separation of concerns
+- Modular design for easy updates and inclusion of new protocols
+- Comprehensive logging and monitoring
+- Simple configuration management
+- Easy troubleshooting and debugging
+
+### Compatibility
+
+- Support for various Valkey deployment models
+- Backward compatibility with existing Valkey features
+- Integration with different LDAP providers
+- Support for multiple authentication methods
+- Compatibility with existing Valkey clients
+
+### High Availabilty
+
+- Multiple identity provider server support
+- Connection failover
+- Cache replication
+- Graceful degradation
+
+### Error Handling
+
+- Invalid credentials
+- Network timeouts
+- LDAP server unavailable
+- Cache corruption
+- Configuration errors
+- Group resolution failures
+
+## Specification
+
+### Authentication flow
+
+1. Client sends AUTH command
+2. Module extracts username/password
+3. Check credential cache
+4. If not cached:
+    - connect to LDAP server
+    - perform bind operation
+    - search for user group membership
+    - cache successful result
+    - set TTL
+5. Pass AUTH/NOAUTH to Valkey core
+6. Return auth response
+
+### On module load
+
+- Load configuration from Valkey config or default values
+- Initialize data structures and connection pools
+- Register command hooks with valkey core
+- Establish initial LDAP server connections
+- Load cached credentials if persistence enabled
+- Register monitoring callbacks
+
+### Commands 
+
+1 LDAP.CONFIG SET <parameter> <value>
+
+- Sets configuration parameters for the LDAP authentication module
+- Parameters include server addresses, timeouts, cache settings, etc.
+- Returns simple string reply: OK on success
+
+2 LDAP.CONFIG GET <parameter>
+
+- Gets current configuration for specified parameter
+- Returns bulk string reply: parameter value
+
+3 LDAP.RELOAD
+
+- Reloads configuration from config file
+- Reestablishes connections with LDAP servers
+- Returns simple string reply: OK on success
+
+4 LDAP.STATUS
+
+- Returns health and statistics information
+- Format: nested hash with connection states, cache stats, auth counts
+
+5 LDAP.FLUSHCACHE
+
+- Clears the credential cache
+- Options to flush all or specific users
+- Returns integer reply: number of entries flushed
+
+### Valkey Command Hooks
+
+#### Authentication Command Hook
+
+- Intercepts AUTH command
+- Processes credentials through LDAP binding
+- Maintains original Valkey auth functionality for local accounts
+
+#### ACL Command Hook
+
+- Integrates with ACL commands
+- Maps LDAP groups to Valkey ACLs
+- Preserves ACL inheritance and evaluation logic
+
+### Internal data structures
+
+#### Configuration Store
+
+- Hash table for configuration parameters
+- Persistent across restarts via RDB or config file
```

Comment:
This is interesting, we don't really do this at all today. I think it was a big mistake in the past to store Function information in the RDB file, but I think storing in the file file seems fine.

Reply:
Valkey stores it in the config file when you do CONFIG REWRITE, doesn't it?

Reply:
`CONFIG REWRITE` doesn't happen on restart though, you have to manually trigger it. Maybe I just was interpreting it a bit differently.

Reply:
You're right, but it's the same with other configs. If we persist the LDAP configs, I think we want to persist the other configs too at the same time, with the same mechanism. An orthogonal problem.